File: secpp/windows/sessman.cpp
// sessman.cpp
#include "sessman.h"

// GuiObject implementation
int GuiObject::curr_handle = 0x120;

GuiObject::GuiObject() : handle(get_handle()) {}

int GuiObject::get_handle() {
    int tmp = GuiObject::curr_handle;
    GuiObject::curr_handle += 4;
    return tmp;
}

// Session implementation
Session::Session(int sess_id) : id(sess_id) {}

std::shared_ptr<Station> Session::new_station(const std::string& name) {
    auto stat = std::make_shared<Station>(name);
    stations.insert({stat->get_handle(), stat});
    return stat;
}

// Station implementation
Station::Station(const std::string& name) : name(name) {}

std::shared_ptr<Desktop> Station::new_desktop(const std::string& name) {
    auto desk = std::make_shared<Desktop>(name);
    desktops.insert({desk->get_handle(), desk});
    return desk;
}

std::string Station::get_name() const {
    return name;
}

// Desktop implementation
Desktop::Desktop(const std::string& name) : name(name) {
    auto win = new_window();
    desktop_window = win;
}

std::shared_ptr<Window> Desktop::new_window() {
    auto window = std::make_shared<Window>();
    windows.insert({window->get_handle(), window});
    return window;
}

std::string Desktop::get_name() const {
    return name;
}

// Window implementation
Window::Window(const std::string& name, const std::string& class_name) {
    this->name = name;
    this->class_name = class_name;
}

// WindowClass implementation
WindowClass::WindowClass(void* class_obj, const std::string& name) : wclass(class_obj), name(name) {}

// SessionManager implementation
SessionManager::SessionManager(const speakeasy::SpeakeasyConfig& cfg)
    : config(cfg), curr_session(nullptr), curr_station(nullptr), curr_desktop(nullptr) {
    dev_ctx = GuiObject::curr_handle;

    // create a session 0
    auto session = std::make_shared<Session>(0);
    sessions.insert({session->get_handle(), session});
    curr_session = session;

    // create WinSta0
    auto st = curr_session->new_station("WinSta0");
    curr_station = st;

    // Create a desktop
    curr_station->new_desktop("Winlogon");
    curr_station->new_desktop("Default");
    curr_station->new_desktop("Disconnect");

    // For now default to the Default desktop
    auto& def_desktops = curr_station->get_desktops();
    for (auto& [hnd, desk] : def_desktops) {
        if (desk->get_name() == "Default") {
            curr_desktop = desk;
            break;
        }
    }
}

int SessionManager::create_window_class(void* class_obj, const std::string& class_name) {
    auto wc = std::make_shared<WindowClass>(class_obj, class_name);
    int atom = wc->get_handle();
    window_classes.insert({atom, wc});
    if (!class_name.empty()) {
        window_classes_by_name.insert({class_name, *wc});
    }
    return atom;
}

int SessionManager::create_window(const std::string& window_name, const std::string& class_name) {
    auto wc = std::make_shared<Window>(window_name, class_name);
    int hnd = wc->get_handle();
    windows.insert({hnd, wc});
    if (!window_name.empty()) {
        windows_by_name.insert({window_name, *wc});
    }
    return hnd;
}

std::shared_ptr<WindowClass> SessionManager::get_window_class(int atom) {
    auto it = window_classes.find(atom);
    if (it != window_classes.end()) {
        return it->second;
    }
    return nullptr;
}

std::shared_ptr<Window> SessionManager::get_window(int handle) {
    auto it = windows.find(handle);
    if (it != windows.end()) {
        return it->second;
    }
    return nullptr;
}

int SessionManager::get_device_context() const {
    return dev_ctx;
}

std::shared_ptr<Desktop> SessionManager::get_current_desktop() {
    return curr_desktop;
}

std::shared_ptr<Station> SessionManager::get_current_station() {
    return curr_station;
}

std::shared_ptr<GuiObject> SessionManager::get_gui_object(int handle) {
    for (auto& sess_pair : sessions) {
        if (sess_pair.first == handle) {
            return sess_pair.second;
        }

        const auto& sess = sess_pair.second;
        for (const auto& stat_pair : sess->get_stations()) {
            if (stat_pair.first == handle) {
                return stat_pair.second;
            }

            for (const auto& desk_pair : stat_pair.second->get_desktops()) {
                if (desk_pair.first == handle) {
                    return desk_pair.second;
                }
            }
        }
    }
    
    return nullptr;
}

File: rust/src/windows/sessman.rs
// Session manager

use std::collections::HashMap;
use std::sync::atomic::{AtomicU32, Ordering};
use crate::config::SpeakeasyConfig;

static CURR_HANDLE: AtomicU32 = AtomicU32::new(0x120);

fn get_next_handle() -> u32 {
    CURR_HANDLE.fetch_add(4, Ordering::SeqCst)
}

#[derive(Debug, Clone)]
pub struct Window {
    pub handle: u32,
    pub name: Option<String>,
    pub class_name: Option<String>,
}

impl Window {
    pub fn new(name: Option<String>, class_name: Option<String>) -> Self {
        Self {
            handle: get_next_handle(),
            name,
            class_name,
        }
    }
}

#[derive(Debug, Clone)]
pub struct WindowClass {
    pub handle: u32,
    pub wclass: u64, // Maps to 'Any' class obj pointer in Python
    pub name: String,
}

impl WindowClass {
    pub fn new(wclass: u64, name: String) -> Self {
        Self {
            handle: get_next_handle(),
            wclass,
            name,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Desktop {
    pub handle: u32,
    pub name: String,
    pub windows: HashMap<u32, Window>,
    pub desktop_window: Window,
}

impl Desktop {
    pub fn new(name: String) -> Self {
        let desk_window = Window::new(None, None);
        let mut windows = HashMap::new();
        windows.insert(desk_window.handle, desk_window.clone());

        Self {
            handle: get_next_handle(),
            name,
            windows,
            desktop_window: desk_window,
        }
    }

    pub fn new_window(&mut self) -> Window {
        let window = Window::new(None, None);
        self.windows.insert(window.handle, window.clone());
        window
    }
}

#[derive(Debug, Clone)]
pub struct Station {
    pub handle: u32,
    pub name: String,
    pub desktops: HashMap<u32, Desktop>,
}

impl Station {
    pub fn new(name: String) -> Self {
        Self {
            handle: get_next_handle(),
            name,
            desktops: HashMap::new(),
        }
    }

    pub fn new_desktop(&mut self, name: &str) -> Desktop {
        let desk = Desktop::new(name.to_string());
        self.desktops.insert(desk.handle, desk.clone());
        desk
    }
}

#[derive(Debug, Clone)]
pub struct Session {
    pub handle: u32,
    pub id: u32,
    pub stations: HashMap<u32, Station>,
}

impl Session {
    pub fn new(sess_id: u32) -> Self {
        Self {
            handle: get_next_handle(),
            id: sess_id,
            stations: HashMap::new(),
        }
    }

    pub fn new_station(&mut self, name: &str) -> Station {
        let stat = Station::new(name.to_string());
        self.stations.insert(stat.handle, stat.clone());
        stat
    }
}

#[derive(Debug, Clone)]
pub enum GuiObject {
    Session(Session),
    Station(Station),
    Desktop(Desktop),
    Window(Window),
    WindowClass(WindowClass),
}

pub struct SessionManager {
    pub sessions: HashMap<u32, Session>,
    pub window_classes_by_handle: HashMap<u32, WindowClass>,
    pub window_classes_by_name: HashMap<String, WindowClass>,
    pub windows_by_handle: HashMap<u32, Window>,
    pub windows_by_name: HashMap<String, Window>,
    pub curr_session: Option<Session>,
    pub curr_station: Option<Station>,
    pub curr_desktop: Option<Desktop>,
    pub dev_ctx: u32,
}

impl SessionManager {
    /// The session manager for the emulator. This will manage things like desktops,
    /// windows, and session isolation
    pub fn new(_config: &SpeakeasyConfig) -> Self {
        let mut sessions = HashMap::new();
        let dev_ctx = get_next_handle();

        // create a session 0
        let mut sess0 = Session::new(0);

        // create WinSta0
        let mut winsta0 = sess0.new_station("WinSta0");

        // Create a desktop
        winsta0.new_desktop("Winlogon");
        let default_desk = winsta0.new_desktop("Default");
        winsta0.new_desktop("Disconnect");

        // Sync updates back to maps
        sess0.stations.insert(winsta0.handle, winsta0.clone());
        sessions.insert(sess0.handle, sess0.clone());

        Self {
            sessions,
            window_classes_by_handle: HashMap::new(),
            window_classes_by_name: HashMap::new(),
            windows_by_handle: HashMap::new(),
            windows_by_name: HashMap::new(),
            curr_session: Some(sess0),
            curr_station: Some(winsta0),
            curr_desktop: Some(default_desk),
            dev_ctx,
        }
    }

    pub fn create_window_class(&mut self, class_obj: u64, class_name: Option<String>) -> u32 {
        let name = class_name.clone().unwrap_or_default();
        let wc = WindowClass::new(class_obj, name.clone());
        let handle = wc.handle;
        
        self.window_classes_by_handle.insert(handle, wc.clone());
        if !name.is_empty() {
            self.window_classes_by_name.insert(name, wc);
        }
        handle
    }

    pub fn create_window(&mut self, window_name: Option<String>, class_name: Option<String>) -> u32 {
        let wc = Window::new(window_name.clone(), class_name);
        let handle = wc.handle;
        
        self.windows_by_handle.insert(handle, wc.clone());
        if let Some(name) = window_name {
            self.windows_by_name.insert(name, wc);
        }
        handle
    }

    pub fn get_window_class_by_handle(&self, handle: u32) -> Option<&WindowClass> {
        self.window_classes_by_handle.get(&handle)
    }
    
    pub fn get_window_class_by_name(&self, name: &str) -> Option<&WindowClass> {
        self.window_classes_by_name.get(name)
    }

    pub fn get_window_by_handle(&self, handle: u32) -> Option<&Window> {
        self.windows_by_handle.get(&handle)
    }
    
    pub fn get_window_by_name(&self, name: &str) -> Option<&Window> {
        self.windows_by_name.get(name)
    }

    pub fn get_device_context(&self) -> u32 {
        self.dev_ctx
    }

    pub fn get_current_desktop(&self) -> Option<&Desktop> {
        self.curr_desktop.as_ref()
    }

    pub fn get_current_station(&self) -> Option<&Station> {
        self.curr_station.as_ref()
    }

    pub fn get_gui_object(&self, handle: u32) -> Option<GuiObject> {
        for (hsess, sess) in &self.sessions {
            if *hsess == handle {
                return Some(GuiObject::Session(sess.clone()));
            }
            for (hstat, stat) in &sess.stations {
                if *hstat == handle {
                    return Some(GuiObject::Station(stat.clone()));
                }
                for (hdesk, desk) in &stat.desktops {
                    if *hdesk == handle {
                        return Some(GuiObject::Desktop(desk.clone()));
                    }
                }
            }
        }
        None
    }
}
